sandbox/mooncake.py

```diff
@@ -0,0 +1,252 @@
+
+    Returns
+    -------
+    sinewave : [batch_size, 100]
+    """
+    sinewaves = []
+    for i in range(batch_size):
+        n = 100  # FIXME np.random.randint(50, 150)
+        f = 1
+        phi = 0
+        x = np.linspace(0, 2 / f, n, dtype=np.float32)
+        sinewave = np.sin(2 * np.pi * f * (x + phi))
+        pad_length = max_num_samples - len(sinewave)
+        sinewave = np.pad(
+            sinewave, (0, pad_length), "constant", constant_values=0)
+        sinewave += 0.1 * np.random.randn(max_num_samples)
+        sinewaves.append(sinewave)
+    return torch.tensor(np.stack(sinewaves, axis=0)).to(DEVICE)
+
+
+def negative_log_prob(x, pi, mu, sigma):
+    """
+    Negative log probability of predictive distribution of x_{k+1} from x_1,
+    ..., x_k.
+
+    Parameters
+    ----------
+    x : [batch_size, max_num_samples]
+        Sinewave.
+    pi, mu, sigma : [batch_size, num_components (+1 for pi), max_num_samples]
+        Mixture weights, means and standard deviations of each Gaussian
+        component, respectively.
+
+    Returns
+    -------
+    [batch_size, max_num_samples - 1]
+    """
+    # Index appropriately to compute nlogp of predicted next value.
+    vals = x.unsqueeze(1)[..., 1:]
+    pi = pi[:, 1:, :-1]
+    mu = mu[..., :-1]
+    sigma = sigma[..., :-1]
+    negative_densities = (0.5 * np.log(2 * np.pi) + torch.log(sigma) -
+                          torch.log(pi) + 0.5 * (vals - mu)**2 / sigma**2)
+    return negative_densities.sum(dim=1)
+    """
+    densities = pi / (sigma * np.log(2*np.pi)) * torch.exp(- (vals - mu)**2 / 2 / sigma**2)
+    return -torch.log(densities.sum(dim=1))
+    """
+
+
+class ResidualBlock(nn.Module):
+    """
+        |-------------------------------------|
+        |                                     |
+        |             |-- tanh --|            |
+    ----|-> dilated_conv         * --- 1x1 -- + -->
+                      |-- sigm --|      |
+                                        |
+                                        |
+    ----------------------------------> + -------->
+    """
+
+    def __init__(self, num_channels, kernel_size, dilation):
+        super(ResidualBlock, self).__init__()
+        self.num_channels = num_channels
+        self.conv1 = CausalConv1d(
+            num_channels,
+            2 * num_channels,
+            kernel_size=kernel_size,
+            dilation=dilation)
+        self.conv2 = nn.Conv1d(
+            num_channels, num_channels, kernel_size=1, dilation=1)
+
+    def forward(self, x):
+        a = self.conv1(x)
+        b = torch.tanh(a[:, :self.num_channels, :])
+        c = torch.sigmoid(a[:, self.num_channels:, :])
+
+        return self.conv2(b * c) + x
+
+
+class Mooncake(nn.Module):
+    # TODO tweak parameters
+    def __init__(
+            self,
+            in_channels=1,
+            num_channels=4,
+            num_blocks=8,
+            kernel_size=2,
+            dilations=[1, 2, 4, 8, 16, 32, 64, 128],
```

Comment:
num_blocks needs to match length of dilations. either error check or make it so it happens automatically